File: src/de/umr/fixcon/Solver.java
package de.umr.fixcon;

import com.google.common.graph.Graphs;

public class Solver {
    Problem problem;
    SubIter iter;
    double globalOptimum;

    public Solver(Problem problem) {
        this.problem = problem;
        iter = new SubIter(problem.graph, problem.subgraphSize);
        globalOptimum = problem.function.optimum(problem.subgraphSize, problem.parameters);
    }

    public Solution getSolution() {
        Solution solution = new Solution(null, Integer.MIN_VALUE);
        while (solution.value < globalOptimum && iter.isValid()) {
            if (problem.function.apply(iter.current(), problem.parameters) > solution.value) {
                solution = new Solution(Graphs.copyOf(iter.current()), problem.function.apply(iter.current(), problem.parameters));
            }
            iter.mutate();
        }
        return solution;
    }
}

File: src/de/umr/fixcon/Solver.kt
package de.umr.fixcon

import de.umr.core.extensions.vertexCount
import de.umr.core.pad
import de.umr.core.removeSmallComponents
import de.umr.fixcon.heuristic.Heuristic
import de.umr.fixcon.itarators.SimpleIter
import de.umr.fixcon.twins.critCliques
import de.umr.fixcon.twins.critIS
import de.umr.fixcon.twins.pruneBigSubsets

fun <V> solve(p: Problem<V>): Solution<V> {
    removeSmallComponents(p.g, p.function.k)

    val sol = Heuristic(p).get()
    if (sol.value == p.function.globalOptimum()) return sol

    val critCliques = critCliques(p)
    critIS(p, critCliques)
    pruneBigSubsets(critCliques, p)

    var iteratorsUsed = 0
    while (sol.value < p.function.globalOptimum() && p.g.vertexCount >= p.function.k) {
        val startVertex = critCliques.subsets().maxBy { it.size }!!.random()
        val iterator = SimpleIter(p, startVertex, sol)

        while (iterator.isValid) iterator.mutate()

        println("BAD VERTICES: ${critCliques[startVertex].size}")
        p.g.removeAllVertices(critCliques[startVertex])
        critCliques.removeSubset(startVertex)
        iteratorsUsed++
    }
    return sol.also { println("Iterators used:".padEnd(pad) + iteratorsUsed) }
}

